### PR TITLE
Normalize composite channels for all user token operations

### LINKED ISSUE #456: [Bug] UserAuthorization refresh sends composite Copilot channel to GetToken
## Summary

`UserAuthorization` normalizes the initial M365 Copilot Web sign-in request to the base `msteams` channel, but its later refresh path sends the raw composite `msteams:COPILOT` channel to Bot Framework `GetToken`.

The initial route auth gate succeeds, then `Authorization.get_token()` / `exchange_token()` refreshes through the raw channel partition and receives HTTP 404. Teams works because its incoming channel is already `msteams`; the failure is specific to M365 Copilot's composite channel.

## Environment

- `microsoft-agents-hosting-core==1.1.0`
- Also reproduced from current `main` at `e24623e`
- Also present in the published `1.2.0.dev9` wheel
- `UserAuthorization` with an Azure Bot OAuth connection
- Custom engine agent invoked from M365 Copilot Web

## Reproduction

1. Configure a normal `UserAuthorization` handler backed by an Azure Bot OAuth connection.
2. Register a message route with that handler in `auth_handlers`.
3. Expose the agent in M365 Copilot Web so inbound activities use `channelId=msteams:COPILOT`.
4. Complete silent SSO or interactive sign-in.
5. Read or exchange the authenticated token in the route with `Authorization.get_token()` or `Authorization.exchange_token()`.

Observed token-service sequence:

```text
GetTokenOrSignInResource(channelId=msteams)          -> 200
GetToken(channelId=msteams:COPILOT)                 -> 404
```

The first request uses the base channel because `#392` calls `get_conversation_reference(force_base_channel=True)` in `_OAuthFlow.begin_flow()`. The refresh request does not.

## Root cause

The remaining path is:

```text
Authorization.exchange_token()
  -> _UserAuthorization.get_refreshed_token()
  -> _UserAuthorization._load_flow()
       flow_state.channel_id = context.activity.channel_id
  -> _OAuthFlow.__init__()
       self._channel_id = flow_state.channel_id
  -> _OAuthFlow.get_user_token()
       UserToken.get_token(channel_id=self._channel_id)
```

`_UserAuthorization._load_flow()` records the raw activity channel in `_FlowState`, and `_OAuthFlow` reuses it for Bot Framework token operations. For Copilot Web this remains `msteams:COPILOT`.

## Expected behavior

Bot Framework token operations should use the base `msteams` channel while OAuth flow storage, continuation activities, telemetry, and replies retain the original `msteams:COPILOT` channel.

The activity itself must not be mutated, because outbound streaming and replies need the original Copilot conversation reference.

## Impact

A user can be signed in successfully at the route gate, but the handler's subsequent token read/refresh fails. The message turn then raises and M365 Copilot displays a generic failure instead of invoking the agent. Repeated attempts continue to fail because the successful base-channel token is queried through the wrong composite partition.

## Proposed fix

Normalize only `_OAuthFlow`'s operational token channel to the base segment while preserving `_FlowState.channel_id` unchanged:

```python
self._channel_id = self._flow_state.channel_id.split(":", 1)[0]
```

This keeps flow storage and continuation semantics surface-specific while making `GetToken`, token exchange, and sign-out consistent with the base-channel behavior already added to `begin_flow()` by `#392`.

A regression test should assert both halves:

- `UserToken.get_token(..., channel_id="msteams")`
- `flow.flow_state.channel_id == "msteams:COPILOT"`

## Validation

An application-level implementation of the same boundary was deployed and verified with consecutive M365 Copilot Web turns:

```text
message_received channel=msteams:COPILOT
 token exchange succeeded
 agent invocation completed successfully
 streamed reply finalized on channel=msteams:COPILOT
```

The previous `GetToken 404` / failed-turn signature disappeared. No tenant IDs, application IDs, tokens, internal URLs, or conversation identifiers are included here.

## Related work

- `#391` described the composite-channel OAuth requirement.
- `#392` fixed the initial `GetTokenOrSignInResource` and encoded-state path.
- `microsoft/Agents-for-net#827` tracks the corresponding composite-channel behavior in .NET.

### PR DESCRIPTION
**Pull request by Carvalh0XYZ**
## TL;DR

Fixes #456.

M365 Copilot custom-engine activities arrive with a composite channel such as `msteams:COPILOT`. Bot Framework user-token storage is partitioned by its base channel (`msteams`). PR #392 normalized only the initial `GetTokenOrSignInResource` path; every other Python `UserToken` operation still sent the composite channel.

In a deployed agent this split caused a deterministic silent-SSO loop:

1. base-channel `GetTokenOrSignInResource` found the resource;
2. `signin/tokenExchange` stored the token under raw `msteams:COPILOT`;
3. replay queried base `msteams`, found no token, and initiated sign-in again;
4. one message produced 10 token-exchange invokes before the M365 `signInInfo` endpoint returned HTTP 429.

This PR follows the maintainer's suggested design and the .NET implementation: normalize at the `UserToken` boundary so every Bot Framework token operation uses the same base-channel partition.

## Change

`UserToken._base_channel_id()` uses `ChannelId(...).channel`, and all six operations normalize before telemetry and HTTP query construction:

- `get_token`
- `_get_token_or_sign_in_resource`
- `get_aad_tokens`
- `exchange_token`
- `get_token_status`
- `sign_out`

This matches .NET `UserTokenRestClient`, which uses `channelId?.Channel` across these operations.

## Safety invariants

- The inbound `Activity` is not mutated.
- OAuth flow state retains the original `msteams:COPILOT` value.
- Replies, continuations, streaming, service URL, and conversation references retain the originating Copilot context.
- Plain channels such as `msteams` are unchanged.
- `None`, empty, whitespace-only, and malformed values are not broadened into an unscoped request.
- Telemetry and HTTP use the same effective channel, matching .NET.

## Regression coverage

The tests use a real local `aiohttp` server and execute all six operations, asserting every wire-level `channelId` is `msteams`. They also cover omitted optional channels and malformed/edge values.

Before this change, the composite-channel test observes `msteams:COPILOT` on those requests; after it, all six use `msteams`.

## Validation

Run from a clean Python 3.12 editable-install environment:

- Focused token/client tests: `139 passed`
- Full repository suite: `2984 passed, 88 skipped`
- Black on changed files: clean
- Focused Flake8: clean
- Independent read-only review: no blockers

## Reviewer guide

1. Review `_base_channel_id()` and the six call sites in `user_token_client.py`.
2. Review the wire-level test covering all six operations.
3. Compare the boundary with .NET `UserTokenRestClient` and the requested direction in #456.

## Related

- #391 / #392 (initial sign-in-resource normalization only)
- microsoft/Agents-for-net#827

### CODE REVIEW

**Comment by Carvalh0XYZ** [comment]
@microsoft-github-policy-service agree

File: tests/hosting_core/_oauth/test_oauth_flow.py [hunk 1]
```diff
@@ -144,6 +144,25 @@ async def test_get_user_token_failure(self, mocker, flow_state):
             code=None,
         )
 
+    @pytest.mark.asyncio
+    async def test_get_user_token_uses_base_channel_for_composite_channel(self, mocker):
+        flow_state = FLOW_DATA.started.model_copy()
+        flow_state.channel_id = "msteams:COPILOT"
+        user_token_client = self.UserTokenClient(
```
[the remark this answers is not in the document]
Comment (Carvalh0XYZ) [reply tests/hosting_core/_oauth/test_oauth_flow.py]: Addressed in ac6aa94. Added composite-channel regression coverage for both sign-out and the signin/tokenExchange continuation. Each test asserts that the Bot Framework token operation receives channel_id="msteams" while flow.flow_state.channel_id remains "msteams:COPILOT". Validation after the change: 137 OAuth-flow tests passed; full suite 2,982 passed and 88 skipped; Black clean across libraries.

**Comment by Carvalh0XYZ** [comment]
Updated per the maintainer suggestion on #456: the implementation now normalizes at the `UserToken` boundary, matching .NET, and covers all six Bot Framework token operations with wire-level regression tests. The activity and OAuth flow state remain untouched. Full suite: 2984 passed, 88 skipped.

File: libraries/microsoft-agents-hosting-core/microsoft_agents/hosting/core/connector/client/user_token_client.py [hunk 2]
```diff
@@ -112,6 +113,15 @@ class UserToken(UserTokenBase):
...
         self.client = client
 
+    @staticmethod
+    def _base_channel_id(channel_id: Optional[str]) -> Optional[str]:
+        """Return the Bot Framework channel without an optional sub-channel."""
+        if not channel_id or not channel_id.strip():
+            return channel_id
+
+        base_channel_id = ChannelId(channel_id).channel
+        return base_channel_id or channel_id
```
[the remark this answers is not in the document]
Comment (Carvalh0XYZ) [reply libraries/microsoft-agents-hosting-core/microsoft_agents/hosting/core/connector/client/user_token_client.py:123]: Addressed in the current PR description and commit 51aa996. The title, change summary, safety invariants, reviewer guide, and regression coverage now explicitly describe normalization at the `UserToken` boundary across all six operations. This broader boundary is intentional and follows the maintainer suggestion on #456 plus the .NET `UserTokenRestClient` design.

**Comment by Carvalh0XYZ** [comment]
Confirmed end-to-end in a deployed M365 Copilot custom-engine agent after resetting both the base and composite Bot Framework token partitions.

### Fresh-user-path result

A brand-new Copilot conversation completed silent SSO with no sign-in card and no client restart:

1. The inbound activity remained `msteams:COPILOT`.
2. `GetTokenOrSignInResource` used `activity.channel_id=msteams` and succeeded.
3. Exactly one `signin/tokenExchange` invoke arrived.
4. `UserToken.exchange_token` used `activity.channel_id=msteams` and succeeded.
5. Continuation replay ran exactly once.
6. `UserToken.get_token` used `activity.channel_id=msteams` and succeeded.
7. OBO, one downstream agent invocation, and Copilot delivery all succeeded.
8. There were no failed spans and no Bot Framework 404 responses.

The original activity, reply context, and delivery telemetry continued to report `msteams:COPILOT`; only Bot Framework token operations used `msteams`.

The M365 client's `signInInfo` request returned HTTP 200 with throttle usage `1/10`. Before normalization, the same clean-token test produced ten token-exchange/continuation attempts, exhausted the quota at `11/10`, and ended in HTTP 429 because exchange stored under `msteams:COPILOT` while replay queried `msteams`.

Our telemetry pipeline duplicates stored span rows, so counts above were deduplicated by operation ID; the runtime console independently shows one `Resending continuation activity` entry.

This validates the boundary requested by the maintainer: normalize all Bot Framework token operations in `UserToken`, while leaving activity and OAuth flow state untouched. PR #457 now implements and tests that shape across all six token operations.

File: libraries/microsoft-agents-hosting-core/microsoft_agents/hosting/core/connector/client/user_token_client.py (hunk 2 above)
Comment (rodrigobr-msft) [inline libraries/microsoft-agents-hosting-core/microsoft_agents/hosting/core/connector/client/user_token_client.py:123]: Small nit: is `or channel_id` needed?

**Review by rodrigobr-msft — COMMENTED** [review]
I will need to perform some testing before approving this, just to check off all the bases, but it looks good.

File: libraries/microsoft-agents-hosting-core/microsoft_agents/hosting/core/connector/client/user_token_client.py (hunk 2 above)
Comment (Carvalh0XYZ) [reply libraries/microsoft-agents-hosting-core/microsoft_agents/hosting/core/connector/client/user_token_client.py:123]: Yes, it handles malformed-but-accepted values like :COPILOT, where .channel is empty, so we preserve the original value instead of accidentally dropping channelId.

**Review by rodrigobr-msft — APPROVED** [review]
Thanks for contributing!
